Google.Api.Gax.Grpc/ClientBuilderBase.cs

```diff
@@ -123,6 +141,104 @@ protected virtual void Validate()
                 TokenAccessMethod, ChannelCredentials);
         }
 
+        /// <summary>
+        /// Performs basic emulator detection and validation based on the given environment variables.
+        /// This method is expected to be called by a derived class that supports emulators, in order to perform the common
+        /// work of checking whether the emulator is configured in the environment.
+        /// </summary>
+        /// <remarks>
+        /// <para>
+        /// If the emulator should not be used, either due to being disabled in <see cref="EmulatorDetection"/> or
+        /// the appropriate environment variables not being set, this method returns null.
+        /// </para>
+        /// <para>
+        /// Otherwise, a dictionary is returned mapping every value in <paramref name="allEmulatorEnvironmentVariables"/> to the value in
+        /// the environment. Any missing, empty or whitespace-only values are mapped to a null reference in the returned dictionary, but
+        /// the entry will still be present (so callers can use an indexer with the returned dictionary for every environment variable passed in).
+        /// </para>
+        /// </remarks>
+        /// <exception cref="InvalidOperationException">The configuration is inconsistent, e.g. due to some environment variables
+        /// being set but not all the required ones, or any environment variables being set in a production-only environment.</exception>
+        /// <param name="requiredEmulatorEnvironmentVariables">Required emulator environment variables.</param>
+        /// <param name="allEmulatorEnvironmentVariables">All emulator environment variables.</param>
+        /// <param name="environmentVariableProvider">The provider used to retrieve environment variables. This is used to faciliate testing, and defaults
+        /// to using <see cref="Environment.GetEnvironmentVariable(string)"/>.</param>
+        /// <returns>A key/value mapping of the emulator environment variables to their values, or null if the emulator should not be used.</returns>
+        protected Dictionary<string, string> GetEmulatorEnvironment(
```

Comment:
proposal (sorry for not thinking of this earlier, and it's fine as it is as well):  change to
`protected bool TryGetEmulatorEnvironment(...., out IDictionary<string, string> environment) { ... }`

Reply:
I think I'd prefer to keep it as it is - because unlike (say) TryParse, it throws exceptions when there are problems.
I think we're fine using null for "don't use an emulator".